src/main/java/com/coveo/spillway/storage/sliding/InMemorySlidingStorage.java

```diff
@@ -0,0 +1,150 @@
+ * buckets, this implementation uses a sliding window of time and counts
+ * the number of requests made in that window.
+ * This implementation CANNOT be used as wrapped storage for {@link AsyncSlidingLimitUsageStorage}.
+ *
+ * @author Emile Fugulin
+ * @since 1.1.0
+ */
+public class InMemorySlidingStorage implements LimitUsageStorage {
+  private Map<Instant, Map<LimitKey, Capacity>> map = new ConcurrentHashMap<>();
+  private Clock clock = Clock.systemDefaultZone();
+  private Duration retention;
+  private Duration slideSize;
+
+  /**
+   * Constructor for the {@link InMemorySlidingStorage}.
+   *
+   * @param retention The retention duration should be greater than highest limit duration, but
+   *                  beware that the higher the retention is, the higher the memory usage will be.
+   * @param slideSize The slide size determines the precision of the count, but beware that a smaller
+   *                  slide size means more memory and more network traffic to sync the limits if needed.
+   */
+  public InMemorySlidingStorage(Duration retention, Duration slideSize) {
+    this.retention = retention;
+    this.slideSize = slideSize;
+  }
+
+  public Duration getSlideSize() {
+    return slideSize;
+  }
+
+  @Override
+  public Map<LimitKey, Integer> addAndGet(Collection<AddAndGetRequest> requests) {
+    Map<LimitKey, Integer> updatedEntries = new LinkedHashMap<>();
+
+    for (AddAndGetRequest request : requests) {
+      Instant bucket =
+          Instant.ofEpochMilli(
+              (request.getEventTimestamp().toEpochMilli() / slideSize.toMillis())
+                  * slideSize.toMillis());
+
+      LimitKey limitKey = LimitKey.fromRequest(request);
+      limitKey.setBucket(bucket);
+
+      Map<LimitKey, Capacity> mapWithThisExpiration =
+          map.computeIfAbsent(bucket, (key) -> new HashMap<>());
+      Capacity counter = mapWithThisExpiration.computeIfAbsent(limitKey, (key) -> new Capacity());
+      counter.addAndGet(request.getCost());
+
+      updatedEntries.put(limitKey, calculateLimitTotal(bucket, request.getExpiration(), limitKey));
+    }
+    removeExpiredEntries();
+
+    return updatedEntries;
+  }
+
+  @Override
+  public Map<LimitKey, Integer> debugCurrentLimitCounters() {
+    removeExpiredEntries();
+    return map.values()
+        .stream()
+        .flatMap(m -> m.entrySet().stream())
+        .collect(Collectors.toMap(Map.Entry::getKey, kvp -> kvp.getValue().get()));
+  }
+
+  @Override
+  public void close() throws Exception {}
+
+  public void applyOnEach(Consumer<Entry<Instant, Map<LimitKey, Capacity>>> action) {
+    map.entrySet().forEach(action);
+    removeExpiredEntries();
+  }
+
+  private void removeExpiredEntries() {
+    Instant oldest = Instant.now(clock).minus(retention);
+    Set<Instant> expiredEntries =
+        map.keySet().stream().filter(bucket -> bucket.isBefore(oldest)).collect(Collectors.toSet());
+    map.keySet().removeAll(expiredEntries);
```

Comment:
isn't this class multithreaded? You'll need synchronization at a bunch of places then or correct data structures...